reana_commons/validation/compute_backends.py

```diff
@@ -0,0 +1,147 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of REANA.
+# Copyright (C) 2022 CERN.
+#
+# REANA is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""REANA-Commons compute backend validation."""
+
+from typing import Dict, List, Optional
+
+from reana_commons.errors import REANAValidationError
+
+
+def build_compute_backends_validator(
+    reana_yaml: Dict, supported_backends: Optional[List[str]]
+):
+    """Validate compute backends in REANA specification file.
+
+    :param reana_yaml: dictionary which represents REANA specification file.
+    :param supported_backends: a list of the supported compute backends.
+    """
+    workflow = reana_yaml["workflow"]
+    workflow_type = workflow["type"]
+    if workflow_type == "serial":
+        workflow_steps = workflow["specification"]["steps"]
+        return SerialComputeBackendValidator(
+            workflow_steps=workflow_steps, supported_backends=supported_backends
+        )
+    if workflow_type == "yadage":
+        workflow_steps = workflow["specification"]["stages"]
+        return YadageComputeBackendValidator(
+            workflow_steps=workflow_steps, supported_backends=supported_backends
+        )
+    if workflow_type == "cwl":
+        workflow_steps = workflow.get("specification", {}).get("$graph", workflow)
+        return CWLComputeBackendValidator(
+            workflow_steps=workflow_steps, supported_backends=supported_backends
+        )
+    if workflow_type == "snakemake":
+        workflow_steps = workflow["specification"]["steps"]
+        return SnakemakeComputeBackendValidator(
+            workflow_steps=workflow_steps, supported_backends=supported_backends
+        )
+
+
+class ComputeBackendValidatorBase:
+    """REANA workflow compute backend validation base class."""
+
+    def __init__(
+        self,
+        workflow_steps: Optional[List[Dict]] = None,
+        supported_backends: Optional[List[str]] = [],
+    ):
+        """Validate compute backends in REANA workflow steps.
+
+        :param workflow_steps: list of dictionaries which represents different steps involved in workflow.
+        :param supported_backends: a list of the supported compute backends.
+        """
+        self.workflow_steps = workflow_steps
+        self.supported_backends = supported_backends
+
+    def validate(self) -> None:
+        """Validate compute backends in REANA workflow."""
+        raise NotImplementedError
+
+    def raise_error(self, compute_backend: str, step_name: str) -> None:
+        """Raise validation error."""
+        raise REANAValidationError(
+            f'Compute backend "{compute_backend}" found in step "{step_name}" is not supported. '
+            f'List of supported compute backends: "{", ".join(self.supported_backends)}"'
+        )
+
+
+class SerialComputeBackendValidator(ComputeBackendValidatorBase):
```

Comment:
Nomenclature comment: if the abstract class is `ComputeBackendValidator`, then specific classes are better named `ComputeBackendValidatorSerial`, `ComputeBackendValidatorYadage`, etc.  This is usually what we use elsewhere, e.g. `JobMonitor` and `JobMonitorKubernetes`, `JobMonitorHTCondor`, etc.  Advantage being the common prefix makes it easy to identify related classes when grepping etc.  

(Kind of like `module.foo`, `module.bar` name association idea.)

Reply:
(Note: We can leave this one for later, as the original code classes were also named like that.)

Reply:
Thanks, changed the naming for all validation classes accordingly